From 41d85b5893e650b40480b9535b24f2f9e3f53e6f Mon Sep 17 00:00:00 2001
From: "Kevin L. Mitchell" <kevin.mitchell@rackspace.com>
Date: Thu, 10 Nov 2011 17:35:44 -0600
Subject: [PATCH] Fix a notification bug when creating instances

Fixes a notifier bug that caused 'launched_at' to be empty and
state to be reported as 'building', due to failure to refresh
instance data prior to generating the notification.  Fixes
bug 834867.

(cherry picked from commit 57e73d374843cade854f07a228192c4f496507ff)

Change-Id: I70fbf3c67407c67b69040ee481ca6d51212cc55d
---
 nova/compute/manager.py    |   12 ++++++------
 nova/tests/test_compute.py |    1 +
 2 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/nova/compute/manager.py b/nova/compute/manager.py
index 1196d98..b926eac 100644
--- a/nova/compute/manager.py
+++ b/nova/compute/manager.py
@@ -459,12 +459,12 @@ class ComputeManager(manager.SchedulerDependentManager):
                                   network_info, block_device_info)
 
             current_power_state = self._get_power_state(context, instance)
-            self._instance_update(context,
-                                  instance_id,
-                                  power_state=current_power_state,
-                                  vm_state=vm_states.ACTIVE,
-                                  task_state=None,
-                                  launched_at=utils.utcnow())
+            instance = self._instance_update(context,
+                                             instance_id,
+                                             power_state=current_power_state,
+                                             vm_state=vm_states.ACTIVE,
+                                             task_state=None,
+                                             launched_at=utils.utcnow())
 
             usage_info = utils.usage_from_instance(instance)
             notifier.notify('compute.%s' % self.host,
diff --git a/nova/tests/test_compute.py b/nova/tests/test_compute.py
index d32c5df..c6747d0 100644
--- a/nova/tests/test_compute.py
+++ b/nova/tests/test_compute.py
@@ -435,6 +435,7 @@ class ComputeTestCase(test.TestCase):
         self.assertTrue('display_name' in payload)
         self.assertTrue('created_at' in payload)
         self.assertTrue('launched_at' in payload)
+        self.assertTrue(payload['launched_at'])
         self.assertEquals(payload['image_ref'], '1')
         self.compute.terminate_instance(self.context, instance_id)
 
-- 
1.7.6.5

